### PR TITLE
Allow async subscriptions to EventStream with dispatchers.

### PR DESCRIPTION
**Pull request by adamhathcock**
Added a null dispatcher

### CODE REVIEW

File: src/Proto.Mailbox/Dispatcher.cs [hunk 1]
```diff
@@ -27,6 +27,17 @@ public interface IDispatcher
...
+        public static NullDispatcher NullDispatcher { get; } = new NullDispatcher();
+    }
+
+    public sealed class NullDispatcher : IDispatcher
+    {
+        public int Throughput => 300;
+
+        public void Schedule(Func<Task> runner)
+        {
+            runner().Wait();
```
Comment (adamhathcock) [inline src/Proto.Mailbox/Dispatcher.cs]: Not sure I like this but seems like the only thing to do.

**Comment by adamhathcock** [comment]
Maybe using `ThreadPoolDispatcher` as the default is better than the "null" one.

File: tests/Proto.Actor.Tests/EventStreamTests.cs [hunk 2]
```diff
@@ -48,5 +49,18 @@ public void EventStream_OnlyReceiveSubscribedToEventTypes()
...
+
+        [Fact]
+        public void EventStream_CanSubscribeToSpecificEventTypes_Async()
+        {
+            var received = "";
+            var eventStream = new EventStream();
+            eventStream.Subscribe<string>(theString =>
+            {
+                received = theString;
+                Assert.Equal("hello", received);
```
Comment (tomliversidge) [inline tests/Proto.Actor.Tests/EventStreamTests.cs]: I don't think this assertion ever gets called - try changing it to "some random string" and it still passes

Comment (tomliversidge) [inline tests/Proto.Actor.Tests/EventStreamTests.cs]: Passing in NullDispatcher will fix it as it runs synchronously

**Comment by tomliversidge** [comment]
Would NullDispatcher be better of named something like ImmediateSynchronousDispatcher or BlockingDispatcher? It's not really null

**Comment by tomliversidge** [comment]
Also, what happens now when a subscriber errors / throws an exception?

**Comment by tomliversidge** [comment]
Like the idea of being able to pass in a dispatcher though! 👍

**Comment by adamhathcock** [comment]
`PassThroughDispatcher` ?

Error handling does need to be thought about, especially now as tasks are being dispatched.  I can do that when people give more thumbs up to this direction.

File: tests/Proto.Actor.Tests/EventStreamTests.cs (hunk 2 above)
Comment (adamhathcock) [inline tests/Proto.Actor.Tests/EventStreamTests.cs]: It does get called, well it did when I was debugging.

Obviously, the test will usually complete before as it's non-deterministic.  Wasn't sure how else to demo things.  By default, it does always use the `NullDispatcher`

Comment (tomliversidge) [inline tests/Proto.Actor.Tests/EventStreamTests.cs]: public static ThreadPoolDispatcher DefaultDispatcher { get; } = new ThreadPoolDispatcher();

- uses ThreadPoolDispatcher by default, hence the test assertion doesn't get called, as the test completes before the assertion.

**Comment by tomliversidge** [comment]
Hmmm.. what is being passed through, and to where? :) The behaviour is "call the function that returns the task , then wait for the task to complete" - 

I find it confusing that the function that returns the task is called "runner" as isn't it the dispatcher that is doing the running? :) - one dispatcher runs things synchronously, the other runs things asynchronously on the threadpool. alternatively you could argue that the threadpool is the runner :) 🥇 

i guess we have "tasks", that get "dispatched" somewhere to be "run"?

**Comment by adamhathcock** [comment]
`SynchronousDispatcher` ?  Naming is hard :)

**Comment by adamhathcock** [comment]
Renamed and added back try/catch in publish.  I hadn't meant to remove it.

**Comment by rogeralsing** [comment]
My main concern for this PR, is that events could be processed out of order as there is no guarantee that tasks that are scheduled on the threadpool run in order.
This is likely fine for many cases, but we might need to add some sort of heads up docs for it

**Comment by adamhathcock** [comment]
I think the best part is that each subscriber can choose.  

I can add to docs...if I knew how :)

**Pull request merged**